### ISSUE TITLE
better spline interpolation

### ISSUE THREAD

**Issue by agriggio** [Beep6581/RawTherapee#4343]
premise: I didn't look into this in detail (yet), these are just user-level observations so far.

The spline interpolation routines that RT uses are sometimes a bit too "unforgiving". What I mean is that often when editing curves in "custom" mode, moving a single point has a big impact on the whole curve. This makes them hard to use, to the point that I almost always prefer a "control cage" curve to the custom one.

But it doesn't have to be this way. A prominent example of a piece of software that does better is GIMP. Here are a couple of quick screenshots to illustrate what I mean:

![rt](https://user-images.githubusercontent.com/11982086/35566417-9c9d06dc-05c1-11e8-8a89-805fbc1eae9b.jpeg)

![gimp](https://user-images.githubusercontent.com/11982086/35566420-a0b86f72-05c1-11e8-82e6-dd2126b37d65.jpeg)

As you can see, the control points are essentially the same, but the GIMP curve is much nicer. In fact, moving the points in GIMP has a much more localized effect than doing the same in RT. I think it's worth looking into this and borrow the GIMP code if possible/feasible. 

(Obviously this is not urgent)

**Comment by raceviper13** [comment]
I'd like this fixed too.  I actually don't like GIMP's either as the radius changes are fairly sudden at the points, but it is more manageable.

These splines could definitely look a whole lot more natural with a curve that has smooth and natural radius changes from end-to-end.

I wish i was a mathematician and could help you out, but i did not do well in my calculus classes.

**Comment by Benitoite** [comment]
It appears that gimp is using a higher-order interpolation..  for example sine^3 vs sine.  I’m not with my graphing calculator at the moment. 🤓

**Comment by agriggio** [comment]
@Benitoite yes I agree, I think they are simply using higher degree splines -- I didn't check though

**Comment by Benitoite** [comment]
It’s a spline called Catmull Rom:
https://github.com/GNOME/gimp/search?utf8=✓&q=+Catmull&type=

**Comment by agriggio** [comment]
Great, thanks for the link! Let's keep this in mind for after 5.4

**Comment by raceviper13** [comment]
I see there is some work in the equation to make sure the curve doesn't double back (the slope should always be from lower left to upper right, and never reverse).  The only part of the GIMP curve that is really weird is the center point.  There seems to be a slope constraint that flattens the curve but there is an equal slope on each side.  it's almost like it's two spline curves joined together.

**Comment by Hombre57** [comment]
Can't  you imagine the control cage in the Gimp example? :wink: I'll try to find a solution on my side too, based on a hidden, computed control cage.

**Comment by Hombre57** [comment]
@agriggio I guess that you're very tempted to copy/pasta the code from Gimp, no problem, but then we should vote which of your solution or mine gives the best result. I'll call mine _Computed Cage_ and will be a new curve type (it won't replace _Custom_).

**Comment by agriggio** [comment]
@Hombre57 I am not tempted to do anything, I just wanted to point out an aspect of the otherwise-excellent curves tool of RT that could be improved imho. do as you please :-)

**Comment by Beep6581** [comment]
Tested branch [`catmull-rom-spline`](https://github.com/Beep6581/RawTherapee/tree/catmull-rom-spline)

That's a nice and easy to use curve @agriggio !

I found these things:
1. When using an auto-matched tone curve, in some cases the new curve leads to this undesirable effect  (see dip in upper-right part) while the old curve was less prone to that, `karin_library_2.hdr.dng` new:
![screenshot_20180726_192744](https://user-images.githubusercontent.com/5844619/43277994-09df2e18-910a-11e8-9bcb-919cb7cd27f8.png)
old:
![screenshot_20180726_192750](https://user-images.githubusercontent.com/5844619/43278002-0e300d0c-910a-11e8-967c-61fa227a99da.png)
or `great_bernera_1.hdr.dng`, new:
![screenshot_20180726_191827](https://user-images.githubusercontent.com/5844619/43277647-f279a1aa-9108-11e8-8df2-46ada7f0f3f7.png)
old:
![screenshot_20180726_191833](https://user-images.githubusercontent.com/5844619/43277656-fbbcad0c-9108-11e8-95d1-51bcf386b9cc.png)
Things are wonky in both cases in `skane_1.hdr.dng`, new:
![screenshot_20180726_193101](https://user-images.githubusercontent.com/5844619/43278146-7d28319e-910a-11e8-8881-a87ddb1bb08e.png)
old:
![screenshot_20180726_193119](https://user-images.githubusercontent.com/5844619/43278150-811b71b2-910a-11e8-8b4c-95c8dfca3879.png)
So the AMTC may need tweaking.
The images are available in http://rawtherapee.com/shared/test_images/
2. Aside from the dip in the upper part of the curve in the previous point, in `great_bernera_2.hdr.dng` the old curve triggered highlight reconstruction to turn on, while the new one does not. I can't say whether it's good or bad - in this case it seems highlight reconstruction doesn't have any visible effect, so maybe it's good. Just observing.

**Comment by agriggio** [comment]
thanks for testing! I'll try to tweak it

**Comment by agriggio** [comment]
@Beep6581 can you try with this patch and see if it improves the situation?

```diff
diff --git a/rtengine/diagonalcurves.cc b/rtengine/diagonalcurves.cc
--- a/rtengine/diagonalcurves.cc
+++ b/rtengine/diagonalcurves.cc
@@ -477,7 +477,7 @@
         if (it+1 < poly_x.end() && t - *it > *(it+1) - t) {
             ++d;
         }
-        return *(poly_y.begin() + d);
+        return LIM01(*(poly_y.begin() + d));
         break;
     }
 
diff --git a/rtengine/histmatching.cc b/rtengine/histmatching.cc
--- a/rtengine/histmatching.cc
+++ b/rtengine/histmatching.cc
@@ -117,8 +117,9 @@
 
     auto coord = [](int v) -> double { return double(v)/255.0; };
     auto doit =
-        [&](int start, int stop, int step, bool addstart) -> void
+        [&](int start, int stop, int step, bool addstart, int maxdelta=0) -> void
         {
+            if (!maxdelta) maxdelta = step * 2;
             int prev = start;
             if (addstart && mapping[start] >= 0) {
                 curve.push_back(coord(start));
@@ -131,7 +132,7 @@
                 }
                 bool change = i > 0 && v != mapping[i-1];
                 int diff = i - prev;
-                if ((change && std::abs(diff - step) <= 1) || diff > step * 2) {
+                if ((change && std::abs(diff - step) <= 1) || diff > maxdelta) {
                     curve.push_back(coord(i));
                     curve.push_back(coord(v));
                     prev = i;
@@ -152,10 +153,11 @@
     int end = mapping.size();
     if (idx <= end / 3) {
         doit(start, idx, idx / 2, true);
-        doit(idx, end, (end - idx) / 3, false);
+        int step = (end - idx) / 3;
+        doit(idx, end, step, false, step);
     } else {
         doit(start, idx, idx > step ? step : idx / 2, true);
-        doit(idx, int(mapping.size()), step, idx - step > step / 2 && std::abs(curve[curve.size()-2] - coord(idx)) > 0.01);
+        doit(idx, end, step, idx - step > step / 2 && std::abs(curve[curve.size()-2] - coord(idx)) > 0.01);
     }
     
     if (curve.size() > 2 && (1 - curve[curve.size()-2] <= step / (256.0 * 3))) {
```

**Comment by Hombre57** [comment]
@agriggio @Beep6581 I know it's too late now, but since this new method replace the `custom` one, we should either add a function in `ProcParams::load` to convert the old curve to the new one by sampling it (e.g.), OR emphasis in the release note that the backward compatibility is seriously broken.

**Comment by agriggio** [comment]
@Hombre57:

https://github.com/Beep6581/RawTherapee/pull/4701#issuecomment-414049370

and following.

TL;DR: I can implement a 'conversion' routine that adds extra control points if needed. if you find examples with old pp3 in which curves are "seriously broken", please let me know and I'll (try to) take care of it
